Automattic/wp-calypso
# Framework: Dump `lib/screen-title` in favor of its Redux counterpart

Or, in more detail:

* Add `getTitle`, `getUnreadCount`, and `getFormattedTitle` selectors (and tests) to `state/document-head`.
* Add `getSectionGroup` selector to `state/ui` (for use by `getFormattedTitle`).
* Remove `document-head` middleware.
* Remove `TitleStore` listener from `client/boot`.
* Instead, `subscribe` a `document.title` setter to the Redux store.
* Replace `lib/screen-title`'s actions with their Redux counterpart (along with a `FIXME` note advising to use `DocumentHead`)
* Remove the `options` parameter from `setTitle` calls. **This is the most critical change here.** Theoretically, `document-head` will fall back to the currently selected site, but since `options` was in some cases used... creatively to set that part of the title to arbitrary strings, this might have broken in cases that want the title to reflect something else (e.g. domains).
* Reader is a bit of a special case because of its  `setPageTitle` helper. I opted to add a `context` arg to that function so it could `context.store.dispatch` the Redux action when called from within a controller; in JSX files, I replaced `this.props.setPageTitle` by `<DocumentHead>` components except for one instance (`FullPost`) where I `connect`ed `setTitle` instead, which seemed easier there.
* Remove `components/data/screen-title`
* Remove `lib/screen-title`

To test:
* In a section you're familiar with, verify that clicking around between different screens updates the title as on `master`. (Maybe open two tabs, one with this branch, and one with wpcalypso, and perform the same actions in both.)
* In mobile mode, check that the `SidebarNavigation` header is updated correctly as well (displaying the correct section title in `/me`  and `/my-sites`).

CC'ing for review:
- [ ] @aduth 
- [ ] @blowery for Reader
- [x] @klimeryk for Domains
- [ ] @scottsweb for VIP

Closes #3796.

Test live: https://calypso.live/?branch=remove/screen-title-flux-lib
